### PR TITLE
feat(tmux): Phase 5 — Bootstrap, Usage Parsing & Cleanup (#180)

### PR DESCRIPTION
**Pull request by dean0x**
## Summary

Phase 5 of the tmux worker migration — persistent session reuse for loops, interactive orchestrator migration to tmux sessions, bootstrap usage parsing, and comprehensive dead code cleanup.

**Problem Being Solved**: Workers run as child processes (piped stdio), which will be metered starting June 15. Tmux sessions avoid this metering model and enable richer agent interaction.

**Key Changes**: 
- EventDrivenWorkerPool uses injected TmuxConnectorPort. Workers identified by tmux session name, not PID.
- Kill sequence is C-c → grace period → force-destroy
- Output routing preserved through existing OutputCapture interface
- Heartbeat detects dead sessions via isAlive()

### Interactive Infrastructure (Steps 1-3)
- TmuxConnector persistent mode with setup shim generation (interactive REPL vs. wrapper pipeline)
- Interactive orchestrator migration from child_process to tmux sessions
- Session name tracking with pid=0 sentinel for tmux workers (migration v30)
- Eager TmuxValidator fail-fast check at bootstrap (tmux >= 3.0 required)

### Persistent Session Reuse (Step 4)
- WorkerPool.persistentSessions map with O(1) lookup by key
- WorkerPool.reuseSession() with AUTOBEAT_TASK_ID env var updates and /clear settle
- LoopHandler sets persistentSessionKey="loop-{loopId}" for single-task iterations
- Cleanup on loop completion, maxIterations, or cancellation
- Edge cases: concurrent reuse guard, dead session fallback, error-driven cleanup
- 10 new tests (worker-pool + loop-handler + domain)

### Dead Code Sweep
- Remove ProcessSpawner dead code from AgentAdapter
- Remove stale spawn/kill/spawnInteractive refs from tests and JSDoc
- Remove orphaned InteractiveSpawnOptions interface
- Remove unused tmuxConnector field from OrchestrationManagerServiceDeps
- Final cleanup of stale timer leak in orchestrate-interactive

## Breaking Changes
- WorkerRegistration.pid is 0 for tmux workers (was worker's actual PID)
- WorkerId format changed from `worker-{pid}` to `worker-beat-{taskId}`
- EventDrivenWorkerPoolDeps interface changed (tmuxConnector + sessionsDir added)
- Requires tmux >= 3.0 installed

## Testing
- 3,459 tests passing across 14 grouped test suites (core, handlers, services, repositories, adapters, implementations, cli, tmux, tmux:integration, dashboard, scheduling, checkpoints, error-scenarios, orchestration)
- All Phase 5 tests: worker-pool spawn/reuse/fallback/cleanup, loop-handler persistence, domain pass-through
- Backward compatibility: pre-Phase-5 orchestrations use SIGTERM fallback

## Related Issues
Closes #180 — Phase 5: Bootstrap Usage Parsing Cleanup
Co-Authored-By: Claude <noreply@anthropic.com>

### CODE REVIEW

**Comment by dean0x** [comment]
**[HIGH - 85% confidence] Persistent session reuse test verifies implementation details, not observable behavior**

`tests/unit/implementations/event-driven-worker-pool.test.ts:868-889`

The "second spawn with same key reuses session" test verifies that `tmuxConnector.spawn` was called once and `setEnvironment`/`sendKeys` were invoked with expected args. However, it doesn't verify the critical observable outcome: the task-to-worker mapping is correctly updated.

```typescript
// Test checks:
expect(tmuxConnector.spawn).toHaveBeenCalledTimes(1);  // Implementation detail
expect(connector.setEnvironment).toHaveBeenCalledWith(...);  // How, not what

// But misses:
// Does pool.getWorkerForTask(task2.id) return the worker? (task re-mapping)
// Does pool.getWorkerForTask(task1.id) return null? (old mapping removed)
```

**Fix**: Add assertions that verify the observable contract:

```typescript
// Verify task re-mapping: task2 is now mapped, task1 is unmapped
const workerForTask2 = pool.getWorkerForTask(task2.id);
expect(workerForTask2.ok).toBe(true);
if (workerForTask2.ok) expect(workerForTask2.value).not.toBeNull();

const workerForTask1 = pool.getWorkerForTask(task1.id);
expect(workerForTask1.ok).toBe(true);
if (workerForTask1.ok) expect(workerForTask1.value).toBeNull();
```

This makes the test more resilient — it validates what `reuseSession()` **does** (re-map tasks) rather than how it does it (calls setEnvironment).

**Note**: The concurrent reuse guard (reuseInProgress set) is also documented but has zero test coverage. Consider adding a test that calls `pool.spawn(task2)` and `pool.spawn(task3)` concurrently with the same persistentSessionKey and verifies one falls back to a fresh spawn.

**Reviewers flagging**: Testing (85% confidence), Behavior-focused test coverage gap

**Comment by dean0x** [comment]
## Code Review: High-Confidence Findings (≥80%)

After analyzing 9 reviewer reports (Architecture, Complexity, Consistency, Reliability, Regression, TypeScript, Performance, Testing, Security), I've identified 5 blocking/critical issues requiring fixes before merge.

### 1. Double `as unknown as` Cast Bypasses Type Safety ⚠️ (85% confidence, HIGH)
**File**: `src/cli/commands/orchestrate-interactive.ts:219-225`

The code casts through `unknown` twice to access `env` on the intentionally-opaque `TmuxSpawnCoreConfig`. This defeats TypeScript's type safety — if the underlying type changes, the cast silently breaks at runtime with no compiler warning.

**Reviewers**: Architecture, Consistency, Complexity, TypeScript

**Fix options**:
1. Add `env` to `TmuxSpawnCoreConfig` (it's already populated at runtime)
2. Use a targeted intersection type instead of `as unknown as`

---

### 2. Orphaned Worker State After reuseSession Failure ⚠️ (90% confidence, HIGH)
**File**: `src/implementations/event-driven-worker-pool.ts:387-396`

Worker state is mutated before the `sendKeys` call. If `sendKeys` fails, `cleanupPersistentSession` destroys the session but does NOT clean up the old `WorkerState` from `this.workers`.

**Consequences**:
- Old worker's heartbeat timer keeps firing, leaking resources
- Destroyed session's `onExit` callback reads the new task ID, stops the new worker's timers, potentially emitting spurious `TaskCompleted` events

**Reviewer**: Reliability

**Fix**: Before returning `ok(null)` from the sendKeys failure path, call `this.cleanupWorkerState(workerId, task.id)` to roll back the state mutation.

---

### 3. Persistent Session Reuse Never Executes in Real Lifecycle ⚠️ (82% confidence, HIGH)
**File**: `src/implementations/event-driven-worker-pool.ts:357-366`

When a loop iteration completes, `onExit` calls `cleanupWorkerState`, removing the worker from `this.workers` map. By the time `reuseSession()` runs for the next task, `this.workers.get(workerId)` returns `undefined` — the reuse path falls through to fresh spawn every time. The persistent session feature is dead code except in tests.

**Reviewer**: Regression

**Fix**: Re-register the worker when existing worker state is missing but the tmux handle is alive. Call `registerWorker()` internally or reconstruct the `WorkerState` from the surviving handle.

---

### 4. Periodic Timers Not Restarted After Session Reuse ⚠️ (82% confidence, MEDIUM)
**File**: `src/implementations/event-driven-worker-pool.ts:310-406`

When `reuseSession()` succeeds, the caller returns immediately without restarting timers that `stopFlushing()` may have cleared:
- `startFlushing()`
- `setupHeartbeatForWorker()`
- `setupTimeoutForWorker()`

Result: Reused session loses output flushing, heartbeat updates, and timeout enforcement.

**Reviewer**: Regression

**Fix**: After successful reuse state updates, restart the three timer functions (contingent on issue #3 being resolved first).

---

### 5. Dead Method: `updateInteractiveOrchestrationPid` ⚠️ (85% confidence, MEDIUM)
**File**: `src/core/interfaces.ts:893`

This method has zero callers. The interactive orchestrator exclusively uses `updateInteractiveOrchestrationSessionName`. The method remains in the interface and implementation as dead weight.

**Reviewer**: Regression

**Fix**: Remove from the `OrchestrationService` interface and its implementation (`orchestration-manager.ts`). This is an internal interface with no external consumers.

---

### Lower-Confidence Findings (60-79%)

These are included in the summary but not as inline comments per the 80% threshold:

- **6 parameters on `spawnAndDeliverPrompt`** (Complexity, 85%) — Consider consolidating into a context object
- **Duplicated finalize+dispose+exit teardown** (Complexity, 82%) — Extract local helper
- **JSDoc line length in `finalizeInteractiveOrchestration`** (Consistency, 82%) — Wrap to 80-90 chars
- **`spawnAndDeliverPrompt` calls `process.exit(1)` but returns `Promise<...|null>`** (Architecture, 82%) — Type lies about nullable return
- **`handleOrchestrateInteractive` is 176+ lines** (Complexity, 82%) — Extract Phase 4 SIGINT block
- **`reuseSession` return type uses null as sentinel** (Consistency, 70%) — Deliberate design choice, documented

---

**Recommendation**: CHANGES_REQUESTED

These issues span reliability (orphaned state), regression (dead code paths and missing timer restarts), and architecture (type safety bypass). All are fixable in one iteration.

**Comment by dean0x** [comment]
## Reliability: Unhandled Process Spawn Errors (Confidence: 80%)

**Lines 326-332** — The tmux attach process only registers an `exit` event handler. If `spawn` fails (e.g., `ENOENT` if tmux vanishes, `EMFILE` on resource exhaustion), the unhandled `error` event can cause the promise to hang indefinitely.

**Fix**: Add an `error` handler to prevent dangling promises:

```typescript
const attachExitCode = await new Promise<number | null>((resolve) => {
  attachProcess.on('exit', (code) => resolve(code));
  attachProcess.on('error', () => resolve(null)); // Treat spawn failure as session-ended
});
```

_Review Cycle 3 — Reliability Reviewer_

File: src/implementations/event-driven-worker-pool.ts [hunk 1]
```diff
@@ -165,8 +190,150 @@ export class EventDrivenWorkerPool implements WorkerPool {
...
+      this.logger.info('Persistent session reused successfully', {
+        taskId: task.id,
+        sessionName: handle.sessionName,
+        key,
+      });
+
+      return ok(existingWorker);
+    } finally {
+      this.reuseInProgress.delete(key);
+    }
```
[the remark this answers is not in the document]
Comment (dean0x) [reply src/implementations/event-driven-worker-pool.ts:471]: **Not an issue.** Callbacks use `taskIdRef.current` (a ref object), not a captured `task.id`. See the `TaskIdRef` pattern at lines 56-66 and the callback implementation at lines 867/882. `reuseSession()` updates `taskIdRef.current` before re-registration (lines 493, 548), so callbacks always route to the current iteration's task ID.

File: src/implementations/tmux/tmux-hooks.ts [hunk 2]
```diff
@@ -155,6 +155,44 @@ exit "$EXIT_CODE"
...
+echo 0 > "$SEQ_FILE"
+
+# Set env vars for the Stop hook.
+export AUTOBEAT_WORKER=true
+export AUTOBEAT_TASK_ID=${singleQuoteToken(config.taskId)}
+export AUTOBEAT_SESSIONS_DIR=${singleQuoteToken(config.sessionsDir)}
+
+# exec-replace this shell with the agent REPL (interactive mode, no output piping).
+# The agent owns the tmux session from this point forward.
+exec ${config.agentCommand} ${agentArgs}
```
[the remark this answers is not in the document]
Comment (dean0x) [reply src/implementations/tmux/tmux-hooks.ts]: **Fixed in f733294.** Both `buildWrapperScript` and `buildSetupShim` now single-quote `agentCommand` via `singleQuoteToken()`, matching the existing quoting on `agentArgs`, `sessionDir`, and all other embedded values.

File: src/implementations/event-driven-worker-pool.ts [hunk 3]
```diff
@@ -165,8 +190,150 @@ export class EventDrivenWorkerPool implements WorkerPool {
...
+          error: clearResult.error.message,
+        });
+        this.cleanupPersistentSession(key);
+        return err(clearResult.error);
+      }
+
+      // Wait for /clear to settle before registering new output handler
+      await new Promise<void>((resolve) => setTimeout(resolve, 300));
+
+      // Re-map the existing worker entry to the new task.
```
[the remark this answers is not in the document]
Comment (dean0x) [reply src/implementations/event-driven-worker-pool.ts]: **Intentional design choice.** Claude Code provides no feedback signal when `/clear` completes — 300ms is the empirically stable minimum. Already documented at lines 147-152 with a note about future DI injection point if needed. Making it configurable adds surface area without solving the root cause (no ready signal).

File: src/implementations/event-driven-worker-pool.ts [hunk 4]
```diff
@@ -159,22 +230,309 @@ export class EventDrivenWorkerPool implements WorkerPool {
...
+        taskIdRef.current = task.id;
+        const regResult = this.registerWorker(task, handle, agentProvider, taskIdRef);
+        if (!regResult.ok) {
+          this.logger.warn('Failed to re-register worker for reused session — spawning fresh', {
+            taskId: task.id,
+            key,
+            error: regResult.error.message,
+          });
+          this.persistentSessions.delete(key);
+          return ok(null);
```
[the remark this answers is not in the document]
Comment (dean0x) [reply src/implementations/event-driven-worker-pool.ts]: **Fixed in f733294.** `reRegisterWorkerForReuse` now calls `cleanupPersistentSession(key)` instead of bare `persistentSessions.delete(key)`, which both removes the map entry AND destroys the tmux session via `tmuxConnector.destroy(handle)`.
